File: Daily-Streak/Day-11_Maximal-Rectangle/Solution.cpp
class Solution {
public:
    // Helper: Largest Rectangle in Histogram
    int largestRectangleArea(vector<int>& heights) {
        stack<int> st;
        int maxArea = 0;
        int n = heights.size();

        for (int i = 0; i <= n; i++) {
            int currHeight = (i == n ? 0 : heights[i]);

            while (!st.empty() && currHeight < heights[st.top()]) {
                int h = heights[st.top()];
                st.pop();
                int width = st.empty() ? i : i - st.top() - 1;
                maxArea = max(maxArea, h * width);
            }
            st.push(i);
        }
        return maxArea;
    }

    int maximalRectangle(vector<vector<char>>& matrix) {
        // ✅ OPTIMAL APPROACH: Histogram + Monotonic Stack
        if (matrix.empty()) return 0;

        int rows = matrix.size();
        int cols = matrix[0].size();
        vector<int> heights(cols, 0);
        int maxRectangle = 0;

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (matrix[i][j] == '1')
                    heights[j] += 1;
                else
                    heights[j] = 0;
            }
            maxRectangle = max(maxRectangle, largestRectangleArea(heights));
        }

        return maxRectangle;
    }

    /*
    ===========================
    ❌ BRUTE FORCE IDEA (COMMENTED)
    ===========================
    - Try all possible rectangles
    - Check if all cells are '1'
    - O(n^3) or worse → TLE

    ===========================
    🧪 STANDALONE VERSION (COMMENTED)
    ===========================
    int main() {
        vector<vector<char>> matrix = {
            {'1','0','1','0','0'},
            {'1','0','1','1','1'},
            {'1','1','1','1','1'},
            {'1','0','0','1','0'}
        };
        Solution s;
        cout << s.maximalRectangle(matrix) << endl;
        return 0;
    }
    */
};

File: Daily-Streak/Day-11_Maximal-Rectangle/Solution.py
class Solution:
    # Helper: Largest Rectangle in Histogram
    def largestRectangleArea(self, heights: List[int]) -> int:
        stack = []
        max_area = 0
        n = len(heights)

        for i in range(n + 1):
            curr_height = 0 if i == n else heights[i]

            while stack and curr_height < heights[stack[-1]]:
                h = heights[stack.pop()]
                width = i if not stack else i - stack[-1] - 1
                max_area = max(max_area, h * width)

            stack.append(i)

        return max_area

    def maximalRectangle(self, matrix: List[List[str]]) -> int:
        # ✅ OPTIMAL APPROACH: Histogram + Monotonic Stack
        if not matrix:
            return 0

        cols = len(matrix[0])
        heights = [0] * cols
        max_rectangle = 0

        for row in matrix:
            for j in range(cols):
                if row[j] == '1':
                    heights[j] += 1
                else:
                    heights[j] = 0

            max_rectangle = max(
                max_rectangle,
                self.largestRectangleArea(heights)
            )

        return max_rectangle

        """
        ===========================
        ❌ BRUTE FORCE IDEA (COMMENTED)
        ===========================
        - Try all rectangles
        - Validate each cell
        - Not feasible

        ===========================
        🧪 STANDALONE VERSION (COMMENTED)
        ===========================
        def main():
            matrix = [
                ['1','0','1','0','0'],
                ['1','0','1','1','1'],
                ['1','1','1','1','1'],
                ['1','0','0','1','0']
            ]
            sol = Solution()
            print(sol.maximalRectangle(matrix))

        if __name__ == "__main__":
            main()
        """
